<!DOCTYPE html>
<html>
  <head>
    <title></title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Working with Genomic Variants
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/Variants_Manipulation_In_R/" class="uri">http://rockefelleruniversity.github.io/Variants_Manipulation_In_R/</a>

---



## Outline
- Manipulate VCF files
- Annotate variants
- Output results

---
## Dataset
&lt;div align="center"&gt;
&lt;img src="imgs/vcfMan_fig3.png" alt="datasetSource" height="400" width="600"&gt;
&lt;/div&gt;

---
## Manipulate VCF files
- Install and load VariantAnnotation package
- Load vcf file 
- General inforamtion in the vcf
- Variants in VRange format
- Information field
- Genotype field

---
## [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html)
&lt;div align="center"&gt;
&lt;img src="imgs/vcfMan_fig2.png" alt="VariantAnnotation" height="400" width="600"&gt;
&lt;/div&gt;

---
## VariantAnnotation ~ load

```r
library(VariantAnnotation)
```

---
## Load vcf file

```r
vcf &lt;- readVcf("data/masterSnpCallsInDels.vcf","hg19")
```

---
## vcf file overview

```r
vcf
```

```
## class: CollapsedVCF 
## dim: 79522 16 
## rowRanges(vcf):
##   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER
## info(vcf):
##   DataFrame with 4 columns: DS, MLEAC, MLEAF, SB
## info(header(vcf)):
##          Number Type    Description                                       
##    DS    0      Flag    Were any of the samples downsampled?              
##    MLEAC A      Integer Maximum likelihood expectation (MLE) for the al...
##    MLEAF A      Float   Maximum likelihood expectation (MLE) for the al...
##    SB    1      Float   Strand Bias                                       
## geno(vcf):
##   SimpleList of length 4: GT, DP, GQ, PL
## geno(header(vcf)):
##       Number Type    Description                                          
##    GT 1      String  Genotype                                             
##    DP 1      Integer Approximate read depth (reads with MQ=255 or with ...
##    GQ 1      Integer Genotype Quality                                     
##    PL G      Integer Normalized, Phred-scaled likelihoods for genotypes...
```

---
## General information

```r
header(vcf)
```

```
## class: VCFHeader 
## samples(16): 1120AJM0001 1120AJM0002 ... 1533AJM0008 1533AJM0009
## meta(4): UnifiedGenotyper fileformat reference contig
## fixed(1): FILTER
## info(4): DS MLEAC MLEAF SB
## geno(4): GT DP GQ PL
```

---
## How many samples in this vcf?

```r
sampleID &lt;- samples(header(vcf))
sampleID
```

```
##  [1] "1120AJM0001" "1120AJM0002" "1120AJM0003" "1316AJM0001" "1316AJM0002"
##  [6] "1371AJM0006" "1371AJM0007" "1371AJM0008" "1371AJM0009" "1371AJM0010"
## [11] "1533AJM0004" "1533AJM0005" "1533AJM0006" "1533AJM0007" "1533AJM0008"
## [16] "1533AJM0009"
```

```r
length(sampleID)
```

```
## [1] 16
```

---
## Metadata field

```r
meta(header(vcf))$META
```

```
## NULL
```

---
## Reference genome

```r
ref_geno &lt;- as.matrix(meta(header(vcf))$reference)
basename(ref_geno[rownames(ref_geno)=="reference"])
```

```
## [1] "Homo_sapiens_assembly19.fasta"
```

---
## Contigs and length

```r
meta(header(vcf))$contig
```

```
## DataFrame with 85 rows and 1 column
##                 length
##            &lt;character&gt;
## 1            249250621
## 2            243199373
## 3            198022430
## 4            191154276
## 5            180915260
## ...                ...
## GL000193.1      189789
## GL000194.1      191469
## GL000225.1      211173
## GL000192.1      547496
## NC_007605       171823
```

---
## Variants information (VRange format)

```r
rd &lt;- rowRanges(vcf)
rd[1:3]
```

```
## GRanges object with 3 ranges and 5 metadata columns:
##               seqnames    ranges strand | paramRangeID            REF
##                  &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;
##   1:69270_A/G        1     69270      * |         &lt;NA&gt;              A
##   1:69511_A/G        1     69511      * |         &lt;NA&gt;              A
##   1:69897_T/C        1     69897      * |         &lt;NA&gt;              T
##                              ALT      QUAL      FILTER
##               &lt;DNAStringSetList&gt; &lt;numeric&gt; &lt;character&gt;
##   1:69270_A/G                  G    734.44           .
##   1:69511_A/G                  G   2219.79           .
##   1:69897_T/C                  C    276.73           .
##   -------
##   seqinfo: 85 sequences from hg19 genome
```

---
## Position of the variations
- chromosome: seqnames()
- start position: start()
- end position: end()

```r
seqnames(rd)[1:2]
```

```
## factor-Rle of length 2 with 1 run
##   Lengths: 2
##   Values : 1
## Levels(85): 1 2 3 4 5 ... GL000194.1 GL000225.1 GL000192.1 NC_007605
```

```r
start(rd)[1:2]
```

```
## [1] 69270 69511
```

```r
end(rd)[1:2]
```

```
## [1] 69270 69511
```
---
## Reference/Alternative Allele
- reference allele is a vector: ref()
- alternative allele is a list: alt()

```r
ref(vcf)[1:2]
```

```
##   A DNAStringSet instance of length 2
##     width seq
## [1]     1 A
## [2]     1 A
```

```r
alt(vcf)[1:2]
```

```
## DNAStringSetList of length 2
## [[1]] G
## [[2]] G
```
---
## Information field ~ Description

```r
info(header(vcf))
```

```
## DataFrame with 4 rows and 3 columns
##            Number        Type
##       &lt;character&gt; &lt;character&gt;
## DS              0        Flag
## MLEAC           A     Integer
## MLEAF           A       Float
## SB              1       Float
##                                                                                                                                                Description
##                                                                                                                                                &lt;character&gt;
## DS                                                                                                                    Were any of the samples downsampled?
## MLEAC    Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed
## MLEAF Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed
## SB                                                                                                                                             Strand Bias
```

---
## Information field

```r
info(vcf)
```

```
## DataFrame with 79522 rows and 4 columns
##                       DS         MLEAC         MLEAF        SB
##                &lt;logical&gt; &lt;IntegerList&gt; &lt;NumericList&gt; &lt;numeric&gt;
## 1:69270_A/G         TRUE            13         0.929  -0.01766
## 1:69511_A/G         TRUE            14           0.7    -90.97
## 1:69897_T/C         TRUE             9           0.9    -22.91
## 1:874816_C/CT      FALSE            NA            NA        NA
## 1:876499_A/G       FALSE            29         0.967    -71.85
## ...                  ...           ...           ...       ...
## Y:15591537_G/C     FALSE             2         0.167    -29.97
## Y:16936081_C/T     FALSE             2         0.143      -208
## Y:16952665_C/T     FALSE             5         0.278      1.43
## Y:21870887_T/G     FALSE             1         0.083    -11.56
## Y:21871374_G/A     FALSE             2         0.167    -96.54
```

---
## Genotype information ~ Description

```r
geno(header(vcf))
```

```
## DataFrame with 4 rows and 3 columns
##         Number        Type
##    &lt;character&gt; &lt;character&gt;
## GT           1      String
## DP           1     Integer
## GQ           1     Integer
## PL           G     Integer
##                                                                               Description
##                                                                               &lt;character&gt;
## GT                                                                               Genotype
## DP              Approximate read depth (reads with MQ=255 or with bad mates are filtered)
## GQ                                                                       Genotype Quality
## PL Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification
```

---
## Genotype information ~ GT

```r
paste0("GT: ",geno(header(vcf))[1,3])
```

```
## [1] "GT: Genotype"
```

```r
matGT &lt;- geno(vcf)$GT
matGT[1:5,1:4]
```

```
##               1120AJM0001 1120AJM0002 1120AJM0003 1316AJM0001
## 1:69270_A/G   "."         "1/1"       "1/1"       "0/0"      
## 1:69511_A/G   "."         "0/1"       "1/1"       "0/0"      
## 1:69897_T/C   "."         "1/1"       "1/1"       "0/0"      
## 1:874816_C/CT "."         "."         "."         "."        
## 1:876499_A/G  "1/1"       "1/1"       "1/1"       "1/1"
```

---
## GT Types

```r
type_GT &lt;- as.vector(as.matrix(matGT))
type_GT[type_GT=="."] &lt;- "NA"
taC &lt;- table(type_GT)
taC
```

```
## type_GT
##    0/0    0/1    0/2    1/1    1/2    1/3    2/2    2/3    3/3     NA 
## 764265 220472    183 204962    114      3    118      1      4  82230
```

---
## GT Types ~ plot

```r
matD &lt;- data.frame(varType=as.vector(names(taC)),
                   Count=as.vector(taC))
ggplot(matD,aes(x=varType,y=Count))+geom_bar(stat='Identity')+
  labs(x="",y="Counts",title="Types of variations")+theme_classic()
```

![](working_with_genomic_variants_files/figure-html/genoGT_varMan_disp-1.png)&lt;!-- --&gt;

---
## Genotype information ~ DP

```r
paste0("DP: ",geno(header(vcf))[2,3])
```

```
## [1] "DP: Approximate read depth (reads with MQ=255 or with bad mates are filtered)"
```

```r
matDP &lt;- geno(vcf)$DP
matDP[1:5,1:4]
```

```
##               1120AJM0001 1120AJM0002 1120AJM0003 1316AJM0001
## 1:69270_A/G            NA         250         250         250
## 1:69511_A/G            NA          71          54         160
## 1:69897_T/C            NA         250         250         250
## 1:874816_C/CT          NA          NA          NA          NA
## 1:876499_A/G            7           5           2           6
```

---
## DP Distribution

```r
dist_DP &lt;- as.vector(as.matrix(matDP))
dist_DP &lt;- dist_DP[!is.na(dist_DP)]
hist(dist_DP,xlab="Depth",main="")
```

![](working_with_genomic_variants_files/figure-html/genoDP_varMan_dist-1.png)&lt;!-- --&gt;

---
## Genotype information ~ GQ

```r
paste0("GQ: ",geno(header(vcf))[3,3])
```

```
## [1] "GQ: Genotype Quality"
```

```r
matGQ &lt;- geno(vcf)$GQ
matGQ[1:5,1:4]
```

```
##               1120AJM0001 1120AJM0002 1120AJM0003 1316AJM0001
## 1:69270_A/G            NA          18          18           9
## 1:69511_A/G            NA          13           9          99
## 1:69897_T/C            NA           6           9           3
## 1:874816_C/CT          NA          NA          NA          NA
## 1:876499_A/G            6           6           3          12
```

---
## GQ Distribution

```r
dist_GQ &lt;- as.vector(as.matrix(matGQ))
dist_GQ &lt;- dist_GQ[!is.na(dist_GQ)]
hist(dist_GQ,xlab="Quality",main="")
```

![](working_with_genomic_variants_files/figure-html/genoGQ_varMan_dist-1.png)&lt;!-- --&gt;

---
## Annotate variantions
- Required resources
- Retrive dbSNP ID
- Predict amino acid changes

---
## Required resources
- [dbSNP](https://bioconductor.org/packages/release/data/annotation/html/SNPlocs.Hsapiens.dbSNP144.GRCh37.html)
- [Reference Genome](http://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html)
- [Annotation](https://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html)


```r
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
```

---
## Predict amino acid changes with TxDb

```r
txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene
```

---
## TxDb

```r
txdb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: GRCh37
## # transcript_nrow: 82960
## # exon_nrow: 289969
## # cds_nrow: 237533
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
## # GenomicFeatures version at creation time: 1.21.30
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
```

---
## Predict amino acid changes ~ Failed

```r
coding &lt;- predictCoding(vcf, txdb, seqSource=Hsapiens)
```
&lt;div align="center"&gt;
&lt;img src="imgs/vcfMan_fig5.png" alt="VariantAnnotation" height="90" width="800"&gt;
&lt;/div&gt;

---
## What's wrong?
- **seqlevels** in the two objects dosn't match
- **seqlevelStyle** in the two objects dosn't match

---
## Check seqlevels

```r
seqlevels(vcf)[1:5]
```

```
## [1] "1" "2" "3" "4" "5"
```

```r
seqlevels(txdb)[1:5]
```

```
## [1] "chr1" "chr2" "chr3" "chr4" "chr5"
```

---
## Change seqlevels

```r
seqlevels(txdb) &lt;- gsub("chr","",seqlevels(txdb))
```

---
## Check seqlevelStyle

```r
seqlevelsStyle(vcf)
```

```
## [1] "NCBI"    "Ensembl"
```

```r
seqlevelsStyle(txdb)
```

```
## [1] "NCBI"    "Ensembl"
```

```r
seqlevelsStyle(Hsapiens)
```

```
## [1] "UCSC"
```

---
## Change seqlevelStyle

```r
seqlevelsStyle(Hsapiens) &lt;- "NCBI"
```

---
## Predict amino acid changes ~ again

```r
coding &lt;- predictCoding(vcf, txdb, seqSource=Hsapiens)
```

```
## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 285132 out-of-bound ranges located on
##   sequences 22, 24, 33, 23, 25, 53, 4115, 4116, 4114, 55, 56, 57,
##   61, 60, 62, 63, 71, 72, 4143, 4145, 4146, 4147, 4148, 4149, 75,
##   76, 77, 78, 79, 4152, 80, 4155, 4158, 4159, 4161, 4156, 4157,
##   4162, 83, 84, 4165, 4166, 4167, 4183, 94, 91, 93, 111, 4193, 4194,
##   4197, 4200, 4201, 120, 4236, 4239, 4240, 121, 4245, 4246, 4247,
##   4248, 4249, 4255, 4256, 135, 136, 137, 138, 4257, 141, 143, 154,
##   155, 156, 157, 160, 161, 163, 4272, 4273, 167, 168, 169, 170, 171,
##   172, 173, 174, 175, 176, 177, 178, 179, 181, 182, 4281, 4284, 190,
##   196, 4285, 202, 203, 4290, 4293, 4298, 4300, 4301, 218, 4318,
##   4319, 4320, 4321, 4323, 4324, 4325, 4326, 4327, 4328, 4329, 4330,
##   4331, 4332, 4333, 4334, 4335, 221, 223, 222, 227, 228, 4341, 4342,
##   4343, 235, 236, 237, 238, 246, 242, 243, 244, 245, 251, 4351,
##   4352, 4353, 4354, 255, 256, 258, 4361, 4364, 4365, 4367, 259, 262,
##   4379, 4380, 4381, 272, 277, 274, 275, 278, 281, 282, 283, 284,
##   285, 286, 287, 288, 289, 291, 296, 294, 299, 300, 303, 301, 302,
##   304, 4390, 4391, 4392, 4394, 4395, 4396, 4397, 4398, 4399, 4400,
##   4406, 4407, 4408, 4413, 336, 337, 338, 4419, 345, 346, 349, 350,
##   351, 354, 358, 359, 361, 366, 367, 369, 370, 371, 372, 4430, 4431,
##   374, 4439, 4442, 390, 391, 392, 393, 389, 403, 417, 418, 419, 420,
##   4447, 4448, 4449, 423, 425, 426, 430, 432, 438, 439, 4464, 440,
##   442, 443, 444, 445, 448, 4471, 4472, 4473, 4477, 4478, 4481, 4482,
##   472, 479, 480, 481, 4513, 4514, 4515, 4516, 4517, 4518, 4519,
##   4520, 4521, 4523, 4524, 4525, 483, 484, 485, 486, 488, 489, 490,
##   491, 493, 494, 495, 496, 498, 499, 500, 503, 509, 510, 511, 4529,
##   4530, 4531, 4532, 4533, 4538, 4539, 4540, 4541, 4542, 4543, 4544,
##   4546, 512, 514, 516, 520, 521, 522, 4564, 4570, 4571, 4572, 4573,
##   4574, 4575, 539, 4581, 4586, 4587, 4588, 4589, 4590, 4592, 4593,
##   4594, 4595, 4596, 4597, 4598, 4601, 4602, 4603, 4604, 4605, 4606,
##   4607, 4608, 4611, 4612, 4613, 4614, 4610, 555, 557, 556, 558, 561,
##   559, 560, 562, 563, 4617, 4618, 4619, 4620, 4621, 564, 4630, 4631,
##   4632, 4633, 566, 567, 568, 581, 582, 583, 585, 586, 588, 589, 590,
##   591, 4651, 4652, 4657, 4658, 4659, 4660, 605, 4671, 4672, 4673,
##   4690, 4691, 4694, 4692, 4693, 4695, 4696, 4697, 4698, 4699, 4700,
##   4702, 4701, 4703, 613, 614, 616, 615, 4707, 4708, 4709, 4710, 632,
##   633, 634, 658, 4739, 674, 675, 676, 681, 682, 683, 684, 685, 689,
##   690, 691, 4754, 4755, 4756, 4758, 4757, 695, 4762, 4763, 4764,
##   4765, 698, 699, 700, 701, 702, 705, 713, 714, 715, 717, 718, 4773,
##   4774, 4776, 4783, 4784, 749, 750, 752, 757, 758, 759, 760, 761,
##   762, 763, 764, 769, 773, 775, 777, 776, 779, 780, 781, 782, 783,
##   784, 786, 791, 792, 793, 799, 800, 801, 803, 809, 810, 4829, 4832,
##   4834, 812, 813, 814, 815, 4837, 4841, 4847, 4848, 4842, 4857, 824,
##   825, 826, 827, 828, 829, 837, 4858, 4859, 4860, 4864, 4865, 4869,
##   4870, 4871, 4866, 4867, 4872, 4879, 4880, 846, 848, 854, 855, 856,
##   857, 858, 859, 860, 4897, 4898, 4899, 4900, 4901, 880, 881, 883,
##   882, 4906, 4907, 885, 886, 887, 4910, 4915, 4916, 4917, 4922,
##   4926, 894, 895, 896, 904, 905, 4929, 4931, 4932, 4933, 4934, 906,
##   907, 908, 4939, 4938, 914, 922, 4954, 4956, 926, 933, 936, 4963,
##   4965, 4964, 4967, 938, 939, 940, 941, 4974, 4975, 942, 944, 948,
##   949, 4976, 953, 954, 955, 957, 958, 959, 960, 961, 962, 4987,
##   4988, 4989, 4993, 4994, 4995, 4998, 4999, 4997, 5000, 5001, 5013,
##   5018, 5030, 5031, 5034, 5035, 5037, 5036, 5043, 5049, 992, 995,
##   989, 990, 991, 999, 1000, 1001, 5061, 5062, 5063, 5065, 5066,
##   5068, 5075, 5078, 5079, 5080, 5082, 5085, 5086, 5087, 5088, 5092,
##   5093, 1022, 1018, 1019, 1020, 1021, 1023, 1024, 5099, 5100, 1029,
##   1030, 1031, 1032, 1028, 1033, 1034, 1035, 1036, 1037, 1038, 1043,
##   1044, 1045, 1046, 1047, 1048, 1049, 1050, 1051, 1052, 1058, 5106,
##   5107, 5108, 5111, 5113, 5115, 5130, 5131, 5132, 5133, 5134, 1064,
##   5135, 5136, 1065, 1068, 1070, 1067, 1077, 1080, 1082, 1083, 5145,
##   5147, 5148, 5149, 1093, 1094, 1095, 1096, 5151, 5152, 1100, 1101,
##   1102, 1103, 1104, 1105, 1106, 1110, 1111, 1112, 5158, 5159, 5160,
##   1117, 1118, 1119, 5161, 5162, 1120, 1121, 1122, 1123, 1127, 1124,
##   1125, 1126, 1136, 1138, 1153, 1164, 1175, 1174, 1177, 1178, 5168,
##   5169, 5170, 5171, 5173, 5174, 1180, 5180, 1199, 1202, 1200, 1201,
##   5184, 5187, 5188, 5189, 5196, 1214, 1215, 1216, 1219, 1220, 5210,
##   5211, 5212, 5213, 1229, 1230, 1231, 1233, 1232, 1235, 5231, 1237,
##   1238, 1240, 1241, 1242, 5236, 5238, 5239, 5241, 1245, 1246, 1247,
##   1250, 1251, 1261, 1262, 5263, 5266, 5267, 1268, 1272, 1273, 1275,
##   5286, 5287, 5288, 5289, 5290, 5291, 5292, 1291, 1292, 1293, 1289,
##   1290, 5297, 5298, 5299, 5300, 5301, 5302, 5303, 5304, 5306, 1295,
##   1296, 1297, 1298, 1299, 1300, 1301, 1303, 1304, 5313, 5314, 5316,
##   5318, 5319, 5322, 5323, 1307, 1308, 1309, 1310, 1311, 1312, 1313,
##   1314, 1315, 1316, 5325, 5337, 1321, 1322, 1323, 5341, 5339, 5342,
##   5343, 5344, 5352, 1331, 1332, 1337, 1339, 1340, 1341, 1346, 1347,
##   1345, 1348, 5363, 5364, 5361, 1350, 5367, 5370, 5378, 5381, 5383,
##   5398, 5400, 5401, 1372, 1373, 1370, 5406, 1375, 1377, 1386, 1390,
##   1391, 5412, 5413, 1393, 1394, 1401, 5422, 5424, 1404, 5426, 5427,
##   5428, 5429, 5430, 5431, 5432, 5433, 5434, 5440, 5442, 1419, 1420,
##   1427, 1428, 1429, 1430, 1433, 1434, 1435, 1436, 1437, 5453, 5454,
##   5455, 5456, 1438, 1439, 1440, 5457, 5458, 5459, 5460, 5461, 5462,
##   1452, 1454, 1455, 1457, 1461, 5473, 5474, 5475, 1471, 1472, 1473,
##   1478, 1479, 1480, 1481, 1482, 1483, 1484, 1485, 1486, 1487, 1491,
##   1496, 5484, 5490, 1512, 1513, 1515, 1522, 1523, 1525, 1527, 1533,
##   1536, 1537, 1544, 1529, 1530, 1546, 1547, 1548, 1550, 1551, 5508,
##   1561, 1562, 1563, 5531, 5532, 5534, 5538, 1585, 1586, 1587, 1588,
##   1589, 1590, 1591, 5540, 5541, 5542, 5543, 5544, 5545, 1597, 1598,
##   5551, 5549, 5547, 5548, 5550, 5552, 5553, 1602, 1603, 1604, 1605,
##   1606, 1607, 1608, 1610, 1618, 1620, 1621, 5557, 5558, 5559, 5560,
##   1625, 5563, 5566, 5568, 5569, 5567, 1631, 1632, 1633, 1634, 5574,
##   1653, 1654, 1655, 1656, 1652, 1658, 1676, 1680, 1681, 1682, 1684,
##   1685, 5587, 5588, 5589, 5594, 5596, 5597, 5605, 1687, 1688, 5611,
##   5612, 5617, 5618, 5624, 5627, 5625, 5630, 5635, 5636, 5637, 1694,
##   1696, 1698, 1699, 1697, 1705, 1706, 1707, 1718, 1719, 1720, 5644,
##   5647, 5648, 5653, 5652, 5655, 5656, 5659, 5661, 5662, 5663, 5664,
##   1724, 1723, 1729, 1734, 1736, 5673, 5675, 5677, 5678, 5679, 5680,
##   5681, 5682, 5683, 5684, 5686, 5685, 1744, 5691, 5692, 5693, 5694,
##   5695, 1759, 1761, 1762, 1767, 1768, 1777, 1778, 1779, 5709, 5710,
##   5714, 5716, 5721, 5722, 5724, 5726, 5727, 1789, 5731, 5732, 1791,
##   1792, 1793, 1794, 1795, 1796, 1797, 1798, 1805, 1808, 1812, 1804,
##   1806, 1809, 1810, 1816, 1817, 1821, 1822, 5743, 5745, 1829, 1835,
##   1837, 1838, 1839, 1840, 1841, 5756, 5757, 5762, 5763, 5764, 1852,
##   1853, 1865, 1866, 1867, 5782, 5783, 5786, 5787, 5788, 5789, 5790,
##   5791, 5796, 5800, 5809, 5810, 5811, 1911, 1912, 1913, 1914, 1915,
##   1916, 1919, 1920, 1921, 5821, 5822, 5823, 5824, 5825, 1929, 1925,
##   1926, 1927, 1928, 1930, 5838, 5840, 1940, 1941, 1952, 1955, 1959,
##   1958, 1960, 1963, 1964, 1965, 1966, 1967, 1968, 1970, 1971, 1972,
##   1974, 5849, 5852, 5853, 5854, 5857, 1976, 1977, 1975, 1978, 1984,
##   1985, 1986, 1991, 1992, 1993, 1994, 5863, 5865, 5874, 5879, 5880,
##   2011, 2012, 2013, 2017, 2018, 2019, 2020, 2022, 2015, 2016, 2021,
##   2023, 2024, 5882, 5883, 5884, 5885, 5892, 5895, 5896, 2035, 2036,
##   2037, 2040, 2043, 2044, 2045, 5902, 5909, 2046, 2048, 2050, 5919,
##   5922, 5923, 5927, 5928, 5929, 5930, 5936, 5947, 5948, 5949, 5950,
##   5952, 5953, 2076, 5957, 5958, 5959, 5960, 5963, 5969, 5970, 5971,
##   5983, 5984, 2088, 2089, 2090, 2092, 2093, 2094, 2096, 2098, 2099,
##   2100, 6003, 6004, 6005, 2105, 2107, 2108, 2109, 2110, 2111, 2112,
##   6008, 6009, 6018, 6021, 6022, 6023, 6025, 6026, 2127, 2128, 2129,
##   2131, 2132, 6030, 2136, 6040, 6041, 6042, 6044, 6047, 6048, 6098,
##   6099, 6096, 6097, 6095, 6101, 6102, 6103, 6104, 6105, 6106, 6087,
##   6089, 2198, 2199, 2200, 2201, 2204, 2238, 2239, 2240, 2241, 2242,
##   2243, 2245, 6121, 6124, 6125, 6154, 6155, 2264, 2269, 2270, 2272,
##   2273, 2274, 2280, 2282, 6189, 6190, 6191, 6272, 6273, 6274, 6283,
##   6284, 62
```

---
## Predict amino acid changes ~ present

```r
coding[1]
```

```
## GRanges object with 1 range and 17 metadata columns:
##               seqnames    ranges strand | paramRangeID            REF
##                  &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt; &lt;DNAStringSet&gt;
##   1:69270_A/G        1     69270      + |         &lt;NA&gt;              A
##                              ALT      QUAL      FILTER      varAllele
##               &lt;DNAStringSetList&gt; &lt;numeric&gt; &lt;character&gt; &lt;DNAStringSet&gt;
##   1:69270_A/G                  G    734.44           .              G
##                  CDSLOC    PROTEINLOC   QUERYID        TXID         CDSID
##               &lt;IRanges&gt; &lt;IntegerList&gt; &lt;integer&gt; &lt;character&gt; &lt;IntegerList&gt;
##   1:69270_A/G       180            60         1           4             4
##                    GENEID CONSEQUENCE       REFCODON       VARCODON
##               &lt;character&gt;    &lt;factor&gt; &lt;DNAStringSet&gt; &lt;DNAStringSet&gt;
##   1:69270_A/G       79501  synonymous            TCA            TCG
##                       REFAA         VARAA
##               &lt;AAStringSet&gt; &lt;AAStringSet&gt;
##   1:69270_A/G             S             S
##   -------
##   seqinfo: 85 sequences from hg19 genome
```

---
## Transform into data frame

```r
alt_base&lt;-NULL;alt_AA&lt;-NULL;prot_posi&lt;-NULL;
for(i in 1:length(names(coding))){
  alt_base &lt;- c(alt_base,paste(as.vector(coding$ALT[[i]]),collapse = ","))
  alt_AA &lt;- c(alt_AA,paste(as.vector(coding$VARAA[[i]]),collapse = ","))
  prot_posi &lt;- c(prot_posi,paste(as.vector(coding$PROTEINLOC[[i]]),collapse = ","))}
#
matA &lt;- data.frame(Variant=names(coding),
                   chromosome=seqnames(coding),
                   start=start(coding),end=end(coding),
                   ref_allele=as.vector(coding$REF),
                   alt_allele=alt_base,
                   GeneID=coding$GENEID,
                   Protein_posi=prot_posi,
                   ref_AA=as.vector(coding$REFAA),
                   alt_AA=alt_AA,
                   Type=coding$CONSEQUENCE)
```

---
## Annotation table ~ Amino Acid Changes

```r
head(matA)
```

```
##         Variant chromosome  start    end ref_allele alt_allele GeneID
## 1   1:69270_A/G          1  69270  69270          A          G  79501
## 2   1:69511_A/G          1  69511  69511          A          G  79501
## 3   1:69897_T/C          1  69897  69897          T          C  79501
## 4 1:874816_C/CT          1 874816 874816          C         CT 148398
## 5 1:874816_C/CT          1 874816 874816          C         CT 148398
## 6 1:874816_C/CT          1 874816 874816          C         CT 148398
##   Protein_posi ref_AA alt_AA          Type
## 1           60      S      S    synonymous
## 2          141      T      A nonsynonymous
## 3          269      S      S    synonymous
## 4          228                  frameshift
## 5          229                  frameshift
## 6          229                  frameshift
```

---
## How many variations in coding region

```r
var_in_coding &lt;- data.frame(varName=names(vcf),
                            in_coding=names(vcf) %in% matA$Variant,
                            stringsAsFactors = FALSE)
table(var_in_coding$in_coding)
```

```
## 
## FALSE  TRUE 
## 19339 60183
```

---
## How many types of variations in coding region

```r
taC &lt;- table(matA$Type)
taC
```

```
## 
##    frameshift      nonsense nonsynonymous    synonymous 
##          1146          1718         79360         76944
```

---
## Variant types in coding region

```r
matD &lt;- data.frame(varType=as.vector(names(taC)),
                   Count=as.vector(taC))
ggplot(matD,aes(x=varType,y=Count))+geom_bar(stat='Identity')+
  labs(x="",y="Counts",title="Types of variations")+theme_classic()
```

![](working_with_genomic_variants_files/figure-html/aaCh_varMan_muType_disp-1.png)&lt;!-- --&gt;

---
## Retrive dbSNP ID

```r
all_snps &lt;- SNPlocs.Hsapiens.dbSNP144.GRCh37
all_snps
```

```
## # SNPlocs object for Homo sapiens (dbSNP Human BUILD 144)
## # reference genome: GRCh37.p13
## # nb of SNPs: 135276726
```

---
## Generate subsets of dbSNPs covering the variations

```r
tar_chr &lt;- as.vector(seqnames(rd)@values)
my_snps &lt;- snpsBySeqname(all_snps,c(tar_chr))
my_snps
```

```
## GPos object with 135274966 positions and 2 metadata columns:
##               seqnames       pos strand |   RefSNP_id alleles_as_ambig
##                  &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; | &lt;character&gt;      &lt;character&gt;
##           [1]        1     10108      * |  rs62651026                Y
##           [2]        1     10109      * | rs376007522                W
##           [3]        1     10139      * | rs368469931                W
##           [4]        1     10150      * | rs371194064                Y
##           [5]        1     10177      * | rs201752861                M
##           ...      ...       ...    ... .         ...              ...
##   [135274962]        Y  59033778      * | rs376130607                Y
##   [135274963]        Y  59033783      * | rs111327212                K
##   [135274964]        Y  59033806      * | rs376828276                K
##   [135274965]        Y  59033957      * | rs371070217                W
##   [135274966]        Y  59033984      * | rs113579233                R
##   -------
##   seqinfo: 25 sequences (1 circular) from GRCh37.p13 genome
```

---
## Annotate Variants in dbSNP

```r
ext_dbSNP &lt;- my_snps[my_snps %in% rd]
```
&lt;div align="center"&gt;
&lt;img src="imgs/vcfMan_fig4.png" alt="VariantAnnotation" height="200" width="600"&gt;
&lt;/div&gt;


---
## Annotate Variants in dbSNP

```r
genome(my_snps) &lt;- "hg19"
ext_dbSNP &lt;- my_snps[my_snps %in% rd]
ext_dbSNP[1:5]
```

```
## GPos object with 5 positions and 2 metadata columns:
##       seqnames       pos strand |   RefSNP_id alleles_as_ambig
##          &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; | &lt;character&gt;      &lt;character&gt;
##   [1]        1     69270      * | rs201219564                R
##   [2]        1     69511      * |   rs2691305                R
##   [3]        1     69511      * |  rs75062661                R
##   [4]        1     69897      * | rs200676709                Y
##   [5]        1    876499      * |   rs4372192                R
##   -------
##   seqinfo: 25 sequences (1 circular) from hg19 genome
```

---
## Generate index for Variant

```r
matV1 &lt;- data.frame(Variant=names(rd))
matV1$posIDX &lt;- gsub("(.*)_(.*)","\\1",matV1$Variant)
head(matV1)
```

```
##         Variant   posIDX
## 1   1:69270_A/G  1:69270
## 2   1:69511_A/G  1:69511
## 3   1:69897_T/C  1:69897
## 4 1:874816_C/CT 1:874816
## 5  1:876499_A/G 1:876499
## 6  1:877831_T/C 1:877831
```

---
## Generate index for SNP_ID

```r
snp_ID &lt;- data.frame(posIDX=paste0(
  seqnames(ext_dbSNP),":",pos(ext_dbSNP)),
                     ext_dbSNP$RefSNP_id)
head(snp_ID)
```

```
##     posIDX ext_dbSNP.RefSNP_id
## 1  1:69270         rs201219564
## 2  1:69511           rs2691305
## 3  1:69511          rs75062661
## 4  1:69897         rs200676709
## 5 1:876499           rs4372192
## 6 1:877831           rs6672356
```

---
## Annotation table ~ SNP_ID

```r
matS &lt;- merge(matV1,snp_ID,all.x=TRUE,
              by.x=2,by.y=1)
matS &lt;- dplyr::select(matS,Variant,ext_dbSNP.RefSNP_id)
names(matS) &lt;- c("Variant","dbSNP_ID")
head(matS)
```

```
##            Variant    dbSNP_ID
## 1  1:100152443_T/C   rs1338576
## 2  1:100154467_G/A   rs6678252
## 3  1:100154899_T/C  rs41285728
## 4 1:100159552_C/CT        &lt;NA&gt;
## 5  1:100195237_G/A rs142004734
## 6  1:100203648_C/T   rs7537296
```

---
## How many variations in dbSNP

```r
taC2 &lt;- table(!is.na(matS$dbSNP_ID))
taC2
```

```
## 
## FALSE  TRUE 
## 11607 68269
```

---
## Variations in dbSNP ~ Plotting

```r
matD &lt;- data.frame(dbSNP=names(taC2),Count=as.vector(taC2))
ggplot(matD,aes(dbSNP,Count))+geom_bar(stat='Identity')+
  labs(x="",y="Counts",title="Variants in dbSNP")+theme_classic()
```

![](working_with_genomic_variants_files/figure-html/dbSNPv_varMan_muCt_disp-1.png)&lt;!-- --&gt;

---
## Integrate annotation information
- Variations
- Amino acid change (matA)
- SNP_ID (matS)
- Information 
  + Genotype (matGT)
  + Depth (matDP)
  + Quality for genotypes (matGQ)

---
## Then..How to extract high-quality variations in a given sample?

Criteria for high-quality variations

- GT: 0/1 (heterozygous) or 1/1 (homozygous)
- DP: greater than or equal to 10
- GQ: greater than or equal to 30

---
## Integrated information 

```r
matC &lt;- data.frame(Variant=names(vcf))
matC &lt;- merge(matC,matA,all.x=TRUE,by=1)
matC &lt;- merge(matC,matS,all.x=TRUE,by=1)
```

---
## Integrated information 

```r
head(matC)
```

```
##            Variant chromosome     start       end ref_allele alt_allele
## 1  1:100152443_T/C       &lt;NA&gt;        NA        NA       &lt;NA&gt;       &lt;NA&gt;
## 2  1:100154467_G/A          1 100154467 100154467          G          A
## 3  1:100154899_T/C          1 100154899 100154899          T          C
## 4 1:100159552_C/CT       &lt;NA&gt;        NA        NA       &lt;NA&gt;       &lt;NA&gt;
## 5  1:100195237_G/A          1 100195237 100195237          G          A
## 6  1:100203648_C/T          1 100203648 100203648          C          T
##   GeneID Protein_posi ref_AA alt_AA          Type    dbSNP_ID
## 1   &lt;NA&gt;         &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;          &lt;NA&gt;   rs1338576
## 2  54873          217      V      V    synonymous   rs6678252
## 3  54873          361      D      D    synonymous  rs41285728
## 4   &lt;NA&gt;         &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;
## 5 391059          276      T      M nonsynonymous rs142004734
## 6 391059          251      Q      Q    synonymous   rs7537296
```

---
## High-quality variations for sample 1

```r
gtP1 &lt;- rownames(matGT)[which(matGT[,1]=="0/1" | matGT[,1]=="1/1")]
dpP1 &lt;- rownames(matDP)[which(matDP[,1]&gt;=10)]
gqP1 &lt;- rownames(matGQ)[which(matGQ[,1]&gt;=30)]
varP1 &lt;- gtP1[which(gtP1 %in% dpP1 &amp; gtP1 %in% gqP1)]
#
length(varP1)
```

```
## [1] 24516
```

---
## High-quality variations for sample 1

```r
matC_P1 &lt;- matC[matC$Variant %in% varP1,]
matC_P1 &lt;- cbind(matC,sampleID=sampleID[1])
head(matC_P1)
```

```
##            Variant chromosome     start       end ref_allele alt_allele
## 1  1:100152443_T/C       &lt;NA&gt;        NA        NA       &lt;NA&gt;       &lt;NA&gt;
## 2  1:100154467_G/A          1 100154467 100154467          G          A
## 3  1:100154899_T/C          1 100154899 100154899          T          C
## 4 1:100159552_C/CT       &lt;NA&gt;        NA        NA       &lt;NA&gt;       &lt;NA&gt;
## 5  1:100195237_G/A          1 100195237 100195237          G          A
## 6  1:100203648_C/T          1 100203648 100203648          C          T
##   GeneID Protein_posi ref_AA alt_AA          Type    dbSNP_ID    sampleID
## 1   &lt;NA&gt;         &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;          &lt;NA&gt;   rs1338576 1120AJM0001
## 2  54873          217      V      V    synonymous   rs6678252 1120AJM0001
## 3  54873          361      D      D    synonymous  rs41285728 1120AJM0001
## 4   &lt;NA&gt;         &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt; 1120AJM0001
## 5 391059          276      T      M nonsynonymous rs142004734 1120AJM0001
## 6 391059          251      Q      Q    synonymous   rs7537296 1120AJM0001
```

---
## Exercises
- [Exercise](./exercise_beta.html)
- [Answer](./answer_beta.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script>
(function() {
  var i, text, code, codes = document.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
})();
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
