---
title: "Genomic Variants ~ Session 3 <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_GenomicVariants/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---
```{r setup_varManS3, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = TRUE,cache.lazy = FALSE,message = FALSE,warning = FALSE, tidy = T)
# AsSlides <- TRUE
#
suppressPackageStartupMessages(library(VariantAnnotation))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(BSgenome.Hsapiens.UCSC.hg19))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
suppressPackageStartupMessages(library(SNPlocs.Hsapiens.dbSNP144.GRCh37))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(GenomicFeatures))
suppressPackageStartupMessages(library(maftools))
suppressPackageStartupMessages(library(NMF))
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# Genomic Variants (part 3)

---
"    
  )
  
}

```

## About this session
In this session, we will demonstrate how to use several advanced analysis approaches to decipher variants from multiple samples. In this situation, the Mutation Annotation Format (MAF) would be a good choice for storing and presenting your data. For a detailesd overview of MAF file structure please refer to our previous course on [Genomic Formats](rockefelleruniversity.github.io/Genomic_Data//presentations/slides/GenomicsData.html#36). 

Today we will beusing an acute myeloid leukemia (AML) dataset (tcga-laml), which was fetched from [The Cancer Genome Atlas data portal](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga). The dataset is in the following path "data/tcga_laml.maf".

---
## Outline of this session
- Handle MAF with maftools
- Get summary by samples and genes
- Any mutation hotspots?
- Any interactions between mutations?
- Any enriched pathways?
- Any Mutational signatures?

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Working with MAF files

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Working with MAF files

---
"    
  )
  
}

```

---
## MAF format
As with many data files we can very simply read MAF files into R.
```{r mult_mafIntro_advan}
laml_tab <- read.delim("data/tcga_laml.maf",sep="\t")
laml_tab[1:2,]
```

---
## How many mutations per sample
```{r mult_samInfo_advan,fig.align="center"}
tbl <- table(laml_tab$Tumor_Sample_Barcode)
hist(tbl,breaks = 10, xlab = "Mutations")
```

---
## [maftools](https://www.bioconductor.org/packages/release/bioc/html/maftools.html)
Bioconductor has a specialist package for handling these files. 
```{r mafTools_intro,echo=FALSE,out.width = "50%",fig.align="center"}
knitr::include_graphics("imgs/vcfMan_fig7r.png")
```

---
## Load MAF file
```{r mult_mafT_advan}
library(maftools)
laml <- read.maf("data/tcga_laml.maf.gz")
class(laml)
```

---
## Summary by samples or genes
maftools provides two functions to generate summary table for by samples or by genes
- getSampleSummary: a matrix, count of mutation classes in each sample
- getGeneSummary: a matrix, count of mutation classes in each gene

---
## How many mutations in each sample
- ***getSampleSummary()***
- a matrix of mutation classes in each sample
```{r mult_samSum_advan}
sample_sum <- getSampleSummary(laml)
sample_sum[1:2,]
```

---
## Mutation classes in each sample
- convert into table for ggplot
- proportion: count of mutation / total mutations for each sample
```{r mult_samSumPlot_advan1}
var_to <- sample_sum$total
names(var_to) <- sample_sum$Tumor_Sample_Barcode
sample_sum <- dplyr::select(sample_sum,-total)
melt_dat <- reshape2::melt(sample_sum, id="Tumor_Sample_Barcode")
melt_dat[1:3,]
```

---
## Mutation classes in each sample
- convert into table for ggplot
- proportion: count of mutation / total mutations for each sample
```{r mult_samSumPlot_advan15}
melt_dat$totalVar <- var_to[match(melt_dat$Tumor_Sample_Barcode,names(var_to))]
melt_dat$prop <- melt_dat$value / melt_dat$totalVar
head(melt_dat)
```

---
## Mutation classes in each sample
```{r mult_samSumPlot_advan2,echo=TRUE,eval=FALSE,tidy=FALSE}
ggplot(melt_dat,aes(x=Tumor_Sample_Barcode,y=value,fill=variable))+
  geom_bar(stat='identity',position = 'stack')+
  labs(x="",y="Mutations",fill="")+
  theme(axis.text.x=element_blank())
```

---
## Mutation classes in each sample
```{r mult_samSumPlot_advan3,echo=FALSE,eval=TRUE,tidy=TRUE,fig.align="center"}
ggplot(melt_dat,aes(x=Tumor_Sample_Barcode,y=value,fill=variable))+
  geom_bar(stat='identity',position = 'stack')+
  labs(x="",y="Mutations",fill="")+
  theme(axis.text.x=element_blank())
```

---
## Mutation classes in each sample
```{r mult_samSumPlotPres_advan4,echo=TRUE,eval=FALSE,tidy=FALSE}
ggplot(melt_dat,aes(x=Tumor_Sample_Barcode,y=prop,fill=variable))+
  geom_bar(stat='identity',position = 'stack')+
  labs(x="",y="Proportion",fill="")+
  theme(axis.text.x=element_blank())
```

---
## Mutation classes in each sample
```{r mult_samSumPlotPres_advan5,echo=FALSE,eval=TRUE,tidy=FALSE,fig.align="center"}
ggplot(melt_dat,aes(x=Tumor_Sample_Barcode,y=prop,fill=variable))+
  geom_bar(stat='identity',position = 'stack')+
  labs(x="",y="Proportion",fill="")+
  theme(axis.text.x=element_blank())
```

---
## How many mutations in each gene
- ***getGeneSummary()***
- A matrix, counts of each mutation classes in each gene
```{r mult_geneSum_advan}
gene_sum <- getGeneSummary(laml)
gene_sum[1:2,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Functional analysis of mutations

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Functional analysis of mutations

---
"    
  )
  
}

```

---
## Any mutation hotspot?
Mutation hotspot means mutation in a given position detected in more samples than in other positions. Most activating mutations have mutation hotspots. A classic example  are the G12, G13, or H61 mutation in KRAS. In contrast, loss of function mutations are usually with no mutation hotspot. The ***lollipop plot*** is a fancy plot of mutations in each position of a given gene. It's very intuitive for demonstrating mutation hotspots.  

---
## Mutation hotspots? 
- ***lollipopPlot()***
```{r mult_lolli_advan_eval1,echo=TRUE,eval=FALSE,tidy=FALSE}
lollipopPlot(maf = laml, 
             gene = 'NRAS', 
             AACol = 'Protein_Change', 
             showMutationRate = TRUE,
             labelPos = "all")
```

---
## Mutation hotspots? 
```{r mult_lolli_advan_eval2,echo=FALSE,eval=TRUE,tidy=FALSE,fig.align="center"}
lollipopPlot(maf = laml, gene = 'NRAS', AACol = 'Protein_Change', showMutationRate = TRUE,labelPos = "all")
```

---
## Interaction between mutations
When mutations are send to interact, this means any concordance or exclusivity of two given gene mutations. If two gene mutations concordantly identified in the same patient, it indicates the two mutations could have synergistic effect in cancer or they are in causality. If two gene mutations are exclusive (i.e. not present in the same patient), this could suggest that they lead to similar consequence, therefore there is no selective pressure for both to mutate. The ***oncoplot*** allows comparison between genes to look for interactions between mutants. 

---
## Oncoplot for top 5 mutated genes
```{r mult_oncoplot_advan,fig.align="center"}
oncoplot(maf=laml,top = 5)
```
**IDH1 and IDH2 mutations could be exlusive.**

---
## Pathway analysis
maftools provides two functions for pathway analysis.
- OncogenicPathways(): calculate mutations detected in each pathway and the fraction of samples affected
- PlotOncogenicPathways(): make a waterfall plot for a given pathway

---
## Any enriched pathways?
```{r mult_pathPlotWF_advan1,eval=TRUE,echo=TRUE,tidy=FALSE,fig.show="hide"}
OncogenicPathways(maf = laml)
```

---
## Any enriched pathways?
```{r mult_pathPlotWF_advan2,eval=TRUE,echo=FALSE,tidy=FALSE,results="hide",fig.align="center"}
OncogenicPathways(maf = laml)
```
**Mutations in RTK-RAS pathway is much higher than others**

---
## Mutations enriched in pathways?
```{r mult_pathPlotWF_advan3,echo=TRUE,eval=TRUE,tidy=FALSE,fig.align="center"}
PlotOncogenicPathways(maf = laml, pathways = "RTK-RAS")
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Mutation Signatures

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Mutation Signatures

---
"    
  )
  
}

```

[Mutational signatures](https://cancer.sanger.ac.uk/cosmic/signatures) mean the patterrns of trinucleotide substitution. As mentioned in [session 1](https://rockefelleruniversity.github.io/RU_GenomicVariants/presentations/slides/RU_GenomicVariant_session1.html#44), the pattern may reflect to exposure of spontaneous/enzymatic deamination and defect of DNA repair, etc.  

This is another common area of study for cancer and genome integrity research.

---
## Steps for mutational signature analysis
- Evaluate nucleotide substitutions
- Calculate trinucleotide matrix
- Estimate Signatures
- Compare to databases
- Compare signature and gene mutations

---
## Ti/Tv plot
- Evaluate single-nucleotide substitution
- ***titv()***
```{r mult_mutSig_TiTv_advan1,eval=FALSE,echo=TRUE,tidy=FALSE}
laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)
plotTiTv(res = laml.titv)
```

---
## Ti/Tv plot
```{r mult_mutSig_TiTv_advan2,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)
plotTiTv(res = laml.titv)
```
**The majority of mutations are C>T.**

---
## Trinucleotide matrix
- calcualte trinucleotide substituion pattern
- ***trinucleotideMatrix()***
```{r mult_mutSig_triMut_advan,tidy=FALSE}
library(BSgenome.Hsapiens.UCSC.hg19, quietly = TRUE)
laml.tnm = trinucleotideMatrix(maf = laml,
                               prefix = 'chr', 
                               add = TRUE, 
                               ref_genome = "BSgenome.Hsapiens.UCSC.hg19")
```

---
## Trinucleotide matrix
```{r mult_mutSig_triMutPres_advan}
dim(laml.tnm$nmf_matrix)
laml.tnm$nmf_matrix[1,]
```

---
## Trinucleotide pattern in TCGA-AB-3009
Lets focus on a single sample
```{r mult_mutSig_triMutPat_advan1}
tarSam_triNuc <- laml.tnm$nmf_matrix['TCGA-AB-3009',]
tarSam_triNuc[1:2]
```

```{r mult_mutSig_triMutPat_advan2}
yd <- data.frame(triNuc=names(tarSam_triNuc),
                 count=tarSam_triNuc,
                 stringsAsFactors = FALSE)
yd$cat <- gsub("(.*)\\[(.*)\\](.*)","\\2",yd$triNuc)
yd$num <- seq(1,length(yd$triNuc))
```

---
## Trinucleotide pattern in TCGA-AB-3009
```{r mult_mutSig_triMutPat_advan3,eval=FALSE,echo=TRUE,tidy=FALSE}
ggplot(yd,aes(x=num,y=count,fill=cat))+
  geom_bar(stat='identity')+
  labs(x="",y="Counts",fill="")+
  theme(axis.text.x=element_blank())
```

---
## Trinucleotide pattern in TCGA-AB-3009
```{r mult_mutSig_triMutPat_advan4,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
ggplot(yd,aes(x=num,y=count,fill=cat))+geom_bar(stat='identity')+
  labs(x="",y="Counts",fill="")+theme(axis.text.x=element_blank())
```
**What does this pattern mean?**

---
## Estimate number of signatures
In this step, [Cophenetic correlation](https://en.wikipedia.org/wiki/Cophenetic_correlation), ***which represents how faithfully a clustering results represent the original data***, would be used to estimate how many signatures could be identified in this dataset. 
```{r mult_mutSig_sigEst_advan1,eval=FALSE,echo=TRUE,tidy=FALSE}
library('NMF')
laml.sign <- estimateSignatures(mat = laml.tnm,
                                nTry = 6,
                                pConstant = 0.1,
                                parallel = 1)
```

---
## Estimate number of signautres
```{r mult_mutSig_sigEst_advan2,eval=TRUE,echo=FALSE,include=FALSE}
library('NMF')
laml.sign <- estimateSignatures(mat = laml.tnm,
                                nTry = 6,
                                pConstant = 0.1,
                                parallel = 1)
```

```{r mult_mutSig_sigEst_advan3,eval=TRUE,echo=FALSE,include=TRUE,fig.align="center"}
plotCophenetic(laml.sign)
```

---
## Extract signautres
- Based on cophenetic plot
- In this case, number of signatures = 3
```{r mult_mutSig_sigExt_advan,tidy=FALSE}
laml.sig.ext <- extractSignatures(mat = laml.tnm, 
                                  n = 3,
                                  pConstant = 0.1,
                                  parallel = 1)
laml.sig.ext$signatures[1:5,] # use for mapping to mutational signature database
```

---
## What do the signtures stand for?
- Map to mutational signatures databases eg. COSMIC
```{r mult_muSig_mapSig_advan,tidy=FALSE}
laml.og30.cosm = compareSignatures(nmfRes = laml.sig.ext,
                                   sig_db = "legacy")
laml.og30.cosm$cosine_similarities[,1:5]
```

---
## Map the signatures to COSMIC database
```{r mult_muSig_mapSigPres_advan,fig.align="center"}
pheatmap::pheatmap(mat = laml.og30.cosm$cosine_similarities, cluster_rows = FALSE)
```

---
## Map the signatures to SBS database
```{r mult_muSig_mapSigSBS_advan1,fig.align="center"}
laml.sign.sbs = compareSignatures(nmfRes = laml.sig.ext, sig_db = "SBS")
laml.sign.sbs$cosine_similarities[,1:5]
```

---
## Map the signatures to SBS database
```{r mult_muSig_mapSigSBS_advan2,fig.align="center"}
pheatmap::pheatmap(mat = laml.sign.sbs$cosine_similarities, cluster_rows = FALSE)
```

---
## plot signatures ~ COSMIC
- plots signature pattern and annotate basing on COSMIC database
```{r mult_muSig_plotSigCOS_advan1,eval=FALSE,echo=TRUE,tidy=FALSE}
plotSignatures(nmfRes = laml.sig.ext, 
               title_size = 1.2,
               contributions = FALSE,
               show_title = TRUE,
               sig_db = 'legacy')
```

---
## plot signatures ~ COSMIC
```{r mult_muSig_plotSigCOS_advan2,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
plotSignatures(nmfRes = laml.sig.ext, 
               title_size = 1.2,
               contributions = FALSE,
               show_title = TRUE,
               sig_db = 'legacy')
```

---
## plot signatures ~ SBS
- plots signature pattern and annotate basing on SBS database
```{r mult_muSig_plotSigSBS_advan1,eval=FALSE,echo=TRUE,tidy=FALSE}
plotSignatures(nmfRes = laml.sig.ext, 
               title_size = 1.2,
               contributions = FALSE,
               show_title = TRUE,
               sig_db = 'SBS')
```

---
## plot signatures ~ SBS
```{r mult_muSig_plotSigSBS_advan2,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
plotSignatures(nmfRes = laml.sig.ext, 
               title_size = 1.2,
               contributions = FALSE,
               show_title = TRUE,
               sig_db = 'SBS')
```


---
## Mutational signautes in each sample
- Display mutational signatures in each sample
```{r mult_muSig_plotSigSAM_advan1,eval=FALSE,echo=TRUE,tidy=FALSE}
plotSignatures(nmfRes = laml.sig.ext,
               title_size = 0.8,
               contributions = TRUE,
               show_title = TRUE)
```

---
## Mutational signautes in each sample
```{r mult_muSig_plotSigSAM_advan2,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
plotSignatures(nmfRes = laml.sig.ext,
               title_size = 0.8,
               contributions = TRUE,
               show_title = TRUE)
```

---
## Enrichment analysis
- k-mean clustering based on the exposure of mutational signatures
- Assign dominate signatures based on k-mean clustering
- Mutations in samples with different signatures
```{r mult_muSig_enrich_advan1,eval=FALSE,echo=TRUE,tidy=FALSE}
laml.se = signatureEnrichment(maf = laml, 
                              sig_res = laml.sig.ext)
```

---
## Enrichment analysis
```{r mult_muSig_enrich_advan2,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center",results=FALSE}
laml.se = signatureEnrichment(maf = laml, sig_res = laml.sig.ext)
```

---
## Genes associated with signatures
- Genes mutated in samples with particular mutational signature
```{r mult_muSig_enrichGene_advan}
laml.se$groupwise_comparision[1:2,]
```

---
## Genes associated with signatures
```{r mult_muSig_enrichGenePlot_advan1,fig.align="center",out.height="70%"}
plotEnrichmentResults(enrich_res = laml.se, pVal = 0.05)
```

---
## Exercise Time
- [Exercise](../../exercises/exercises/GenomicVariant_session3_exercise.html)
- [Answer](../../exercises/answers/GenomicVariant_session3_answers.html)