---
title: "Genomic Variants ~ Session 2 <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_GenomicVariants/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r setup_varManS2, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = TRUE,cache.lazy = FALSE)
# AsSlides <- TRUE
#
suppressPackageStartupMessages(library(VariantAnnotation))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(BSgenome.Hsapiens.UCSC.hg19))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
suppressPackageStartupMessages(library(SNPlocs.Hsapiens.dbSNP144.GRCh37))
suppressPackageStartupMessages(library(ggplot2))
vcf <- readVcf("data/SAMN01882168_filt.vcf.gz","hg19")
#vcf <- readVcf("../data/SAMN01882168_filt.vcf.gz","hg19")
rd <- rowRanges(vcf)
```

# Abou this session
In this session, we are focus on **variant annotation**. You will learn how to annotate variants with rsID of dbSNP and consequences of amino acid changes by [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html). The VCF file is the same as session 1. You can find it in this file path "data/SAMN01882168_filt.vcf.gz". Besides, the annotation process is highly memory intensive. So, we just focus on the variants in Chromosome 1. 

Required resources
- dbSNP: [SNPlocs.Hsapiens.dbSNP144.GRCh37](https://bioconductor.org/packages/release/data/annotation/html/SNPlocs.Hsapiens.dbSNP144.GRCh37.html)
- Annotation databse: [TxDb.Hsapiens.UCSC.hg19.knownGene](https://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html)

```{r annoRS_varMan}
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
```

---
# Outline of this session
- Subset variants in chromosome 1
- Annotate rsID from dbSNP
- Predict amino acid changes
- Integrate information into single table

---
# Subsetting Variants in chromosome 1
```{r chrSub,eval=TRUE,tidy=FALSE,echo=TRUE}
vcf_chr1 <- vcf[grepl(names(vcf),pattern = "chr1:")]
rd_chr1 <- rowRanges(vcf_chr1)
```

---
# Annotate rsID from dbSNP
- Retrieve dbSNP
- Extract SNPs by Chromosome ***memory intensive***
- Merge SNPs and variant information by positions

---
## Retrive dbSNP data

```{r dbSNPv_varMan}
all_snps <- SNPlocs.Hsapiens.dbSNP144.GRCh37
all_snps
```

---
## Retrieve SNPs by chromosome
- Extract chromosome name from the VRange object: ***seqnames()@values***
- Retrive SNPs by chromosome: ***snpsBySeqname()***
```{r dbSNPv_varMan_load_fail,eval=FALSE}
tar_chr <- as.vector(seqnames(rd_chr1)@values)
my_snps <- snpsBySeqname(all_snps,c(tar_chr))
```
```{r dbSNPv_varMan_load_fail2,eval=TRUE,echo=FALSE,fig.align="center",out.width="75%"}
knitr::include_graphics("imgs/vcfMan_fig5r.png")
```

---
## What's wrong?
- **seqlevels** in the two objects dosn't match
- **seqlevelStyle** in the two objects dosn't match
- **genome** in the tow objects dosen't match

---
## check seqlevels
```{r check_seqLvl}
# seqlevels ~ rd
seqlevels(rd_chr1)
# seqlevels ~ dbSNP
seqlevels(all_snps)
```
seqname in rd_chr1 is *chr1* but in all_snps is *1*

---
## check seqlevelStyle
```{r check_seqStyle}
seqlevelsStyle(rd_chr1)
seqlevelsStyle(all_snps)
```
seqlevelsSytle in rd_chr1 is *UCSC* but in all_snps is *NCBI Ensembl*

---
## check genome
```{r check_genome}
genome(rd_chr1)
genome(all_snps)
```
genome in rd_chr1 is *hg19* but in all_snps is *GRCh37.p13*

---
## Unify seqnames and Process
```{r dbSNPv_varMan_load}
tar_chr <- as.vector(seqnames(rd_chr1)@values)
tar_chr <- gsub("chr","",tar_chr)
tar_chr[grepl(tar_chr,pattern = "M")] <- "MT"
my_snps <- snpsBySeqname(all_snps,c(tar_chr))
my_snps[1:2]
```

---
## Replace seq Info of my_snps
- set seqlevels to "chr" and change MT to M
- set seqlevelsStyle  to "UCSC"
- set genome to "hg19"
```{r change_seqlvl}
# change seqlevels
seqlevels(my_snps) <- paste0("chr",seqlevels(my_snps))
seqlevels(my_snps) <- gsub("MT","M",seqlevels(my_snps))
# change seqlevelsStyle
seqlevelsStyle(my_snps) <- "UCSC"
# change genome
genome(my_snps) <- "hg19"
```

---
## Make rsID table
- posIDX: [chromosome]:[position] ~ the index column for merging data
- rsID: rsID from dbSNP 
```{r makeTab_varMan}
snp_ID <- data.frame(
  posIDX=paste0(seqnames(my_snps),":",pos(my_snps)),
  rsID=my_snps$RefSNP_id,stringsAsFactors = FALSE)
head(snp_ID)
```

---
## Generate Variant table
```{r dbSNPV_varMan_tbl_1}
matV1 <- data.frame(Variant=names(rd_chr1),stringsAsFactors = FALSE)
matV1$chromosome <- gsub("(.*):(.*)_(.*)/(.*)","\\1",matV1$Variant)
matV1$start <- gsub("(.*):(.*)_(.*)/(.*)","\\2",matV1$Variant)
matV1$end <- gsub("(.*):(.*)_(.*)/(.*)","\\2",matV1$Variant)
matV1$ref_allele <- gsub("(.*):(.*)_(.*)/(.*)","\\3",matV1$Variant)
matV1$alt_allele <- gsub("(.*):(.*)_(.*)/(.*)","\\4",matV1$Variant)
matV1$posIDX <- gsub("(.*)_(.*)","\\1",matV1$Variant)
matV1[1:2,]
```

---
## Annotation table ~ SNP_ID
```{r dbSNPV_varMan_tbl_3,tidy=FALSE}
matS <- merge(matV1,snp_ID,all.x=TRUE,by="posIDX")
matS <- dplyr::select(matS,-posIDX)
matS[1:2,]
```

---
## How many variations in dbSNP
```{r dbSNPV_varMan_muCt,tidy=FALSE}
taC2 <- table(!is.na(matS$rsID))
taC2_dat <- as.data.frame(taC2)
taC2
```

---
## Variations in dbSNP ~ Plotting
```{r dbSNPv_varMan_muCt_disp1,tidy=FALSE,eval=FALSE,echo=TRUE}
ggplot(taC2_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='Identity')+
  labs(x="",y="Counts",fill="in_dbSNP")+
  theme(legend.position = "none")
```

---
## Variations in dbSNP ~ Plotting
```{r dbSNPv_varMan_muCt_disp2,tidy=FALSE,echo=FALSE,eval=TRUE,fig.align="center"}
ggplot(taC2_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='Identity')+
  labs(x="",y="Counts",fill="in_dbSNP")+
  theme(legend.position = "none")
```

---
# Predict amino acid changes
- load annotation database: TxDb
- predict amino acid changes: ***predictCoding***

---
## Load TxDb
```{r aaCh_varMan}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```

---
## TxDb
```{r aaCh_varMan_txdb}
txdb
```

---
## Predict amino acid changes
```{r aaCh_varMan_pred}
coding <- predictCoding(vcf_chr1, txdb, seqSource=Hsapiens)
```

---
## Variants and predicted consequence
```{r aaCh_varMan_pres2}
coding[1]
```

---
## Transform into data frame
```{r aaCh_varMan_frame,eval=TRUE,tidy=FALSE,echo=TRUE}
matA <- data.frame(Variant=names(coding),
                   chromosome=seqnames(coding),
                   start=start(coding),end=end(coding),
                   ref_allele=as.character(coding$REF),
                   alt_allele=unlist(lapply(lapply(
                     coding$ALT,`[[`,1),as.character)),
                   GeneID=coding$GENEID,
                   TxID=coding$TXID,
                   Protein_posi=unlist(lapply(lapply(
                     coding$PROTEINLOC,`[[`,1),as.integer)),
                   ref_AA=as.character(coding$REFAA),
                   alt_AA=as.character(coding$VARAA),
                   Type=coding$CONSEQUENCE,
                   stringsAsFactors = FALSE)
matA$aaChange <- paste0("p.",matA$ref_AA,matA$Protein_posi,matA$alt_AA)
matA <- dplyr::select(matA,-Protein_posi,-ref_AA,-alt_AA)
```

---
## Annotation table ~ Amino Acid Changes
```{r aaCh_varMan_tbl}
matA[1:2,]
```

---
## How many variations in coding region
```{r aaCh_varMan_muCt}
var_in_coding <- data.frame(varName=names(vcf_chr1),
                            in_coding=names(vcf_chr1) %in% matA$Variant,
                            stringsAsFactors = FALSE)
table(var_in_coding$in_coding)
```

---
## How many types of mutations in coding region
```{r aaCh_varMan_muType}
taC <- table(matA$Type)
taC_dat <- as.data.frame(taC)
taC
```
- nonsense: mutations causing the appearance of stop codon
- nonsynonymous: mutations causing amino acid changes
- synonymous: mutations not causing amino acid changes

---
## Mutation types in coding region
```{r aaCh_varMan_muType_disp1,tidy=FALSE,echo=TRUE,eval=FALSE}
ggplot(taC_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='Identity')+
  labs(x="",y="Counts",fill="")+
  theme(legend.position = "none")
```

---
## Mutation types in coding region
```{r aaCh_varMan_muType_disp2,tidy=FALSE,echo=FALSE,eval=TRUE,fig.align="center"}
ggplot(taC_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='Identity')+
  labs(x="",y="Counts",fill="")+
  theme(legend.position = "none")
```

---
# Integrate SNP and amino acid change into single table
```{r comb_varMan}
matS$GeneID <- matA$GeneID[match(matS$Variant,matA$Variant)]
matS$AAChange <- matA$GeneID[match(matS$Variant,matA$Variant)]
matS[1:2,]
```

---
# Other annotation softwares
- [ANNOVAR](https://doc-openbio.readthedocs.io/projects/annovar/en/latest/)
  + Wang et al, NAR, 2010(38):e164 [link](https://academic.oup.com/nar/article/38/16/e164/1749458)
  + Detailed databases available for human genome
  + Perl-based

- [SnpEff](http://snpeff.sourceforge.net/)
  + Cingolani et al, Landes Biosciense, 2012(6):1 [link](http://snpeff.sourceforge.net/SnpEff_paper.pdf)
  + Support over 38,000 genomes
  + Java-based

---
# Exercise Time
- [Exercise](../../exercises/exercises/GenomicVariant_session2_exercise.html)
- [Answer](../../exercises/answers/GenomicVariant_session2_answers.html)