---
title: "Genomic Variants ~ Session 1 <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_GenomicVariants/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---
```{r setup_varMan, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = TRUE,cache.lazy = FALSE, tidy = T)
# AsSlides <- TRUE
#
suppressPackageStartupMessages(library(VariantAnnotation))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(BSgenome.Hsapiens.UCSC.hg19))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
suppressPackageStartupMessages(library(SNPlocs.Hsapiens.dbSNP144.GRCh37))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(GenomicFeatures))
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# Genomic Variants (part 1)

---
"    
  )
  
}

```

## About this session
Variant Call Format (VCF) is widely used file format to record genomic variants. We introduced the VCF file format in [previous section](rockefelleruniversity.github.io/Genomic_Data//presentations/slides/GenomicsData.html#35). In this session, we want to demonstrate how to manipulate VCF file with [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html).

We will be using a study on bladder cancer from Wei Wangs group at BGI-Shenzhen as our dataset in this session:
- Raw data from SRA: [PRJNA185252](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA185252) [LINK](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3696090/)
- Reference genome: [BSgenome.Hsapiens.UCSC.hg19](https://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html)
- Variant calling procedure is based on [GATK Best Practice](https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-)
  + Align to reference genome: [Burrows-Wheeler Aligner](http://bio-bwa.sourceforge.net/bwa.shtml), bwa-mem
  + Sort and remove duplicates: [Picard-tools](http://broadinstitute.github.io/picard/command-line-overview.html), [SortSam](https://broadinstitute.github.io/picard/command-line-overview.html#SortSam)/[MarkDuplicates](https://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates)
  + Recalibrate Base Quality Scores: [GATK](https://gatk.broadinstitute.org/hc/en-us/articles/360036433132--Tool-Documentation-Index), [BaseRecalibrator](https://gatk.broadinstitute.org/hc/en-us/articles/360036434532-BaseRecalibrator)/[ApplyBQSR](https://gatk.broadinstitute.org/hc/en-us/articles/360036433652-ApplyBQSR)
  + Variant Calling: [GATK](https://gatk.broadinstitute.org/hc/en-us/articles/360036433132--Tool-Documentation-Index), [HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/360036803991-HaplotypeCaller)/[GenotypeGVCFs](https://gatk.broadinstitute.org/hc/en-us/articles/360036806011-GenotypeGVCFs)
- Demo data: SAMN01882168 ("data/SAMN01882168_filt.vcf.gz")

---
## Outline of this session
- Load VCF file into a vcf object
- Survey and retrieve information from each field
  + META field: general information of the VCF file
  + Fixed field: variants, usually in VRange format
  + Genotype field: information for each sample
  + Integrate information into single table
- Differentiate variant types
- Retrieve nucleotide substitution pattern
- Trinucleotide motif analysis

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# VCF files

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# VCF files

---
"    
  )
  
}

```


## [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html)
```{r varLoad_intro,echo=FALSE,out.width = "75%",fig.align="center"}
knitr::include_graphics("imgs/vcfMan_fig2.png")
```

---
## Load VCF file
```{r varLoad_varMan}
library(VariantAnnotation)
vcf <- readVcf("data/SAMN01882168_filt.vcf.gz","hg19")
vcf
```

---
## What is in each field?
- META field: general information of the VCF file

- FIX field: variants, usually in VRange format

- GENOTYPE field: information for each sample

- Integrate information into single table

---
## Overview
As with objects we have met in other courses, there are accessor functions to grab the contents of our VCF object. 
```{r gInfo_varMan}
header(vcf)
```
- META field: meta
- FIX field: rowRange, info
- GENOTYPE field: geno

---
## How many samples are in this VCF?
```{r sample_varMan}
sampleID <- samples(header(vcf))
sampleID
```

---
## What is in the META field?
```{r metaOV_META_varMan}
meta(header(vcf))
```

---
## Retrieve information in the META field
Once we have the META extracted, we can use the dollar sign to extract specific fields. 
```{r meta_META_varMan}
# File format
meta(header(vcf))$fileformat
# Source used for variant calling
meta(header(vcf))$source
```

---
## Retrieve information in the META field
Once we have the META extracted, we can use the dollar sign to extract specific fields. 
```{r meta_contig_varMan}
meta(header(vcf))$contig
```

---
## Variants information (VRange format)
The variant information is recorded in a VRange object. This is a form of [GRanges](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/slides/GenomicIntervals_In_Bioconductor.html#6) object which we regularly use and are integral to Bioconductor. It includes chromosome, start position, end potision, reference allele, and alternative allele.
```{r range_varMan}
rd <- rowRanges(vcf)
rd[1:2]
```
***It's not a strand-specific library. This is indictaed by the \*, which means unstranded. ***

---
## Retrieving variation information
We can use similar accessors to those used for [GRanges](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/slides/GenomicIntervals_In_Bioconductor.html#16) to get position information. 
- chromosome: seqnames()
- start position: start()
- end position: end()

---
## Position of the variations
```{r range_varMan_posi}
as.vector(seqnames(rd)[1:2]) # Chromosome
start(rd)[1:2] # Start position
end(rd)[1:2] # End position
```

---
## How to get the reference alleles?
- The *ref()* function can be used to get the reference allele for a variant. The result is a DNAStringSet.
```{r range_varMan_baseInfo_Ref1}
refBase <- ref(vcf)
refBase[1:2]
```
- Use as.character to covert it to character
```{r range_varMan_baseInfo_Ref3}
refBase <- as.character(refBase)
refBase[1:2]
```

---
## How to get alternative alleles?
The *alt()* function can be used to get the alternative allele for a variant. The result is a DNAStringSetList ***as alternative alleles could be multiple changes.***
```{r range_varMan_baseInfo_Alt1}
altBase <- alt(vcf)
alt(vcf)[1:2] 
```

---
## How to get alternative alleles?
- As it is a list, getting a character back is more complicated. First we can extract vectors from list by using lapply(). Specifically at this point we are subsetting the data to only the first variant. 
```{r range_varMan_baseInfo_Alt3}
# get the 1st vector of the list
altBase <- lapply(altBase,`[[`,1) 
altBase[1:2]
```
- Convert DNAString object to character
```{r range_varMan_baseInfo_Alt4}
altBase <- unlist(lapply(altBase,as.character)) 
altBase[1:2]
```

---
# INFO section from FIXED field
- **Integrated information from all the samples**
- Annotation information would be recorded in this section
- All information is stored as a ***data frame***
```{r info_varMan}
info(header(vcf))[1:2,]
```

---
## INFO section
By calling *info()* directly on the whole VCF object, we get information for every variant.
```{r info_varMan_disp}
info(vcf)[1:2,]
```

---
# GENOTYPE field
- Separated by individual samples
```{r geno_varMan}
geno(header(vcf))[1:2,]
```

---
## Genotype information for each sample
By calling *geno()* directly on the whole VCF object, we get information for every variant.
```{r genoGT_varMan}
paste0("GT: ",geno(header(vcf))[1,3])
matGT <- geno(vcf)$GT
matGT[1:2,]
```

---
## GT Types
```{r genoGT_varMan_tbl}
tbl <- table(geno(vcf)$GT)
tbl_dat <- as.data.frame(tbl)
tbl
```
- 0/1: ***heterozygous mutations, one allele is the same as reference sequence***
- 1/1: ***homozygous mutations, both alleles are different from reference sequence***
- 1/2: ***heterozygous mutations, both alleles are different from reference sequence***

---
## GT Types
```{r genoGT_varMan_disp1,echo=TRUE,tidy=FALSE,eval=FALSE}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='Identity')+
  labs(x="",y="Counts",fill="")+
  theme_classic()
```

---
## GT Types
```{r genoGT_varMan_disp2,echo=FALSE,tidy=FALSE,eval=TRUE,fig.align="center"}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='Identity')+
  labs(x="",y="Counts",fill="")+theme_classic()
```

---
## Depth for each sample (DP)
```{r genoDP_varMan}
paste0("DP: ",geno(header(vcf))[3,3])
matDP <- geno(vcf)$DP
matDP[1:2,]
```

---
## DP Distribution
```{r genoDP_varMan_dist1,eval=FALSE,echo=TRUE,tidy=FALSE}
summary(as.vector(matDP))
```

```{r genoDP_varMan_dist2,eval=TRUE,echo=FALSE,results=TRUE}
summary(as.vector(matDP))
```

```{r genoDP_varMan_distPres1,fig.align="center",eval=FALSE,echo=TRUE,tidy=FALSE}
ggplot(as.data.frame(matDP),aes(x=SAMN01882168))+geom_histogram()+
  labs(x="",y="Counts")+
  scale_x_log10()+
  theme_classic()
```

---
## DP Distribution
```{r genoDP_varMan_distPres2,fig.align="center",eval=TRUE,echo=FALSE}
ggplot(as.data.frame(matDP),aes(x=SAMN01882168))+geom_histogram()+
  labs(x="",y="Counts")+
  scale_x_log10()+
  theme_classic()
```

---
## Genotype calling quality (GQ)
- The confidence in genotype calling
- Expressed as [Phred quality score](https://gatk.broadinstitute.org/hc/en-us/articles/360035531872-Phred-scaled-quality-scores)
```{r genoGQ_varMan}
paste0("GQ: ",geno(header(vcf))[4,3])
matGQ <- geno(vcf)$GQ
matGQ[1:2,]
```

---
## GQ distribution
```{r genoGQ_varMan_dist1,eval=FALSE,echo=TRUE,tidy=FALSE}
summary(as.vector(matGQ))
```

```{r genoGQ_varMan_dist2,eval=TRUE,echo=FALSE,tidy=FALSE,results=TRUE}
summary(as.vector(matGQ))
```

```{r genoGQ_varMan_distPres1,fig.align="center",eval=FALSE,echo=TRUE,tidy=FALSE}
ggplot(as.data.frame(matGQ),aes(x=SAMN01882168))+geom_histogram()+
  labs(x="",y="Counts")+
  scale_x_log10()+
  theme_classic()
```

---
## GQ Distribution
```{r genoGQ_varMan_distPres2,fig.align="center",eval=TRUE,echo=FALSE,warning=FALSE}
ggplot(as.data.frame(matGQ),aes(x=SAMN01882168))+geom_histogram()+
  labs(x="",y="Counts")+scale_x_log10()+theme_classic()
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Gathering variant information

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Gathering variant information

---
"    
  )
  
}

```

## Gathering information ~ GT 0/1 and 1/1
- select variants with GT 0/1 or 1/1
```{r gatGT_info1,echo=TRUE,tidy=FALSE}
var_1 <- rownames(geno(vcf)$GT)[
  geno(vcf)$GT=="0/1" | 
    geno(vcf)$GT=="1/1"]
```
- Extract variant information
```{r gatGT_info2,echo=TRUE,tidy=FALSE}
varTab1 <- data.frame(variant=names(rd)[names(rd) %in% var_1],
                      chr=as.vector(seqnames(rd)[names(rd) %in% var_1]),
                      start=start(rd)[names(rd) %in% var_1],
                      end=end(rd)[names(rd) %in% var_1],
                      stringsAsFactors = FALSE)
```

---
## Gathering information ~ GT 0/1 and 1/1
- Ref alleles are retrieved from ref(vcf)
```{r gatGT_info3,echo=TRUE,tidy=FALSE}
ref_base <- ref(vcf)[rownames(vcf) %in% var_1]
ref_base[1:2]
varTab1$refBase <- as.character(ref_base)
```

---
## Gathering information ~ GT 0/1 and 1/1
- Alt alleles are retrieved from alt(vcf)
```{r gatGT_info4,echo=TRUE,tidy=FALSE}
alt_base <- lapply(alt(vcf)[rownames(vcf) %in% var_1],`[[`,1)
alt_base[1]
alt_base <- lapply(alt_base,as.character)
alt_base[1]
varTab1$altBase <- unlist(alt_base)
```

---
## Gathering information ~ GT 0/1 and 1/1
- Extract counts from AD
```{r gatGT_info5,echo=TRUE,tidy=FALSE}
adCount <- geno(vcf)$AD[rownames(geno(vcf)$AD) %in% var_1]
adCount[1]
varTab1$refCount <- unlist(lapply(adCount,`[[`,1))
varTab1$altCount <- unlist(lapply(adCount,`[[`,2))
```

---
## Gathering information ~ GT 0/1 and 1/1
- genoType: genotype (GT)
- gtQuality: genotyping quality (GQ)
```{r gatGT_info6,echo=TRUE,tidy=FALSE}
varTab1$genoType <- geno(vcf)$GT[rownames(geno(vcf)$GT) %in% var_1]
varTab1$gtQuality <- geno(vcf)$GQ[rownames(geno(vcf)$GQ) %in% var_1]
```

---
## Gathering information ~ genotype 1/2
- Heterozygous mutations, both alleles are different from the reference sequence.
- Both ref allele and alt allele have to be retrieved from alt(vcf)
```{r gatGT_info7,echo=TRUE,tidy=FALSE}
var_2 <- rownames(geno(vcf)$GT)[geno(vcf)$GT=="1/2"]
varTab2 <- data.frame(variant=names(rd)[names(rd) %in% var_2],
                      chr=as.vector(seqnames(rd)[names(rd) %in% var_2]),
                      start=start(rd)[names(rd) %in% var_2],
                      end=end(rd)[names(rd) %in% var_2],
                      refBase=unlist(lapply(lapply(
                        alt(vcf)[rownames(vcf) %in% var_2],`[[`,1),as.character)),
                      altBase=unlist(lapply(lapply(
                        alt(vcf)[rownames(vcf) %in% var_2],`[[`,2),as.character)),
                      refCount=unlist(lapply(
                        geno(vcf)$AD[rownames(geno(vcf)$AD) %in% var_2],`[[`,2)),
                      altCount=unlist(
                        lapply(geno(vcf)$AD[rownames(geno(vcf)$AD) %in% var_2],`[[`,3)),
                      genoType=geno(vcf)$GT[rownames(geno(vcf)$GT) %in% var_2],
                      gtQuality=geno(vcf)$GQ[rownames(geno(vcf)$GQ) %in% var_2],
                      stringsAsFactors = FALSE)
```

---
## Merged datatable
```{r gatGT_merge}
varTab <- rbind(varTab1,varTab2)
varTab[1:2,]
```

---
## Differentiate variant types
- SNP: single-nucleotide substitutions
- DEL: deletions
- INS: insertions
- Others: complicated variations, such as Ins/Del or Inversion

---
## Variant types
```{r mutType_gen}
# differentiate SNP/INS/DEL/Others
for(k in 1:length(varTab$variant)){
  if(width(varTab$refBase[k]) < width(varTab$altBase[k])){
    varTab$mutType[k] <- "INS"
  }else if(width(varTab$refBase[k]) > width(varTab$altBase[k])){
    varTab$mutType[k] <- "DEL"
  }else if(width(varTab$refBase[k])==1&width(varTab$altBase[k])==1){
    varTab$mutType[k] <- "SNP"
  }else{
    varTab$mutType[k] <- "Others"}}
#
tbl <- table(varTab$mutType)
tbl_dat <- as.data.frame(tbl)
tbl
```

---
## Variant types
```{r mutType_pres1,echo=TRUE,eval=FALSE,tidy=FALSE,fig.align="center"}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat = 'identity')+
  labs(x="",y="Mutations",fill="")+
  theme_classic()
```

---
## Variant types
```{r mutType_pres2,echo=FALSE,eval=TRUE,fig.align="center"}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+geom_bar(stat = 'identity')+
  labs(x="",y="Mutations",fill="")+theme_classic()
```

---
## Nucleotide substitution pattern
- only SNPs
- Transition (Ti): purine-to-purine, pyrimidine-to-pyrimidine
- Transversion (Tv): purine-to-pyrimidine, pyrimidine-to-purine

---
## Nucleotide substitution
```{r TiTb_gen,tidy=FALSE}
# Transition (Ti)
ti <- c("A>G","G>A","C>T","T>C")
# Transveersion (Tv)
tv <- c("A>T","A>C","G>T","G>C","C>A","C>G","T>A","T>G")
varTab$nuSub <- paste0(varTab$refBase,">",varTab$altBase)
varTab$TiTv[varTab$nuSub %in% ti] <- "Ti"
varTab$TiTv[varTab$nuSub %in% tv] <- "Tv"
varTab[1:2,]
```

---
## Nucleotide substitution
```{r TiTv_pres_nuSub}
varX <- varTab[varTab$mutType=="SNP",]
tbl <- table(varX$nuSub)
tbl_dat <- as.data.frame(tbl)
tbl
```

---
## Nucleotide substitution
```{r TiTv_pres_nuSubPres1,eval=FALSE,tidy=FALSE,echo=TRUE,fig.align="center"}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat = 'identity')+
  labs(x="",y="Mutations",fill="")+
  theme(legend.position = "none")
```

---
## Nucleotide substitution
```{r TiTv_pres_nuSubPres2,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+geom_bar(stat = 'identity')+
  labs(x="",y="Mutations",fill="")+theme(legend.position = "none")
```

---
## Ti/Tv
```{r TiTv_pres_TiTv1}
tbl <- table(varX$TiTv)
tbl_dat <- as.data.frame(tbl)
tbl
```

---
## Ti/Tv
```{r TiTv_pres_TiTv2,echo=TRUE,eval=FALSE,tidy=FALSE}
ggplot(as.data.frame(table(varX$TiTv)),aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat = 'identity')+labs(x="",y="Mutations",fill="")+
  theme(legend.position = "none")
```

---
## Ti/Tv
```{r TiTv_pres_TiTv=3,echo=FALSE,eval=TRUE,tidy=FALSE,fig.align="center"}
ggplot(as.data.frame(table(varX$TiTv)),aes(x=Var1,y=Freq,fill=Var1))+geom_bar(stat = 'identity')+labs(x="",y="Mutations",fill="")+theme(legend.position = "none")
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Trinucleotide motif analysis

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Trinucleotide motif analysis

---
"    
  )
  
}

```


## Trinucleotide motif analysis
Other constraints may bias nucleotide transitions. One of these is the trinucleotide motif around the variant (from -1 to +1). This pattern could reflect different causes for the mutations.  

One such example is APOBEC, a deaminase leading to C>T substitutions in the motif ***T[C>T]W***. So, the proportion of ***T[C>T]W*** in overall C>T substitutions could reflect to APOBEC activity. More applications of the trinucleotide motifs will be discussed in [session3](https://rockefelleruniversity.github.io/RU_GenomicVariants/presentations/slides/RU_GenomicVariant_session3.html#19).

---
## Extract C>T substituion
```{r motif_load_advAn}
library(BSgenome.Hsapiens.UCSC.hg19)
library(GenomicFeatures)
library(stringr)
#
rd_idx<-str_split(names(rd),"_", simplify=T)
rd_idx[1:2,]
rd_sub <- rd[rd_idx[,2]=="C/T"]
rd_sub[1:2,]
```

---
## Extract sequences beneath the mutation from -1 to +1
We can use the **getSeq()** function to get the sequence underlying these variants, as we have done with [GRanges](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/slides/GenomicIntervals_In_Bioconductor.html#63).
```{r motif_seqExt_advAn,tidy=FALSE}
rd_sub$triNu <- getSeq(Hsapiens,
                 seqnames(rd_sub),
                 start=start(rd_sub)-1,
                 end=end(rd_sub)+1)
rd_sub[1:2]
```

---
## Trinucleotide pattern
```{r motif_seqPat_advan}
tbl <- table(rd_sub$triNu)
tbl_dat <- as.data.frame(tbl)
tbl
```

---
## Trinucleotide pattern
```{r motif_seqPat_advan2,eval=FALSE,echo=TRUE,tidy=FALSE}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='identity')+
  labs(x="",y="Variants",fill="")+
  theme(legend.position = "none")
```

---
## Trinucleotide pattern
```{r motif_seqPat_advan3,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
ggplot(tbl_dat,aes(x=Var1,y=Freq,fill=Var1))+
  geom_bar(stat='identity')+
  labs(x="",y="Variants",fill="")+
  theme(legend.position = "none")
```

---
## APOBEC activation in this sample?
- APOBEC target: TCW (TCA/TCT)
```{r motif_ApoTar_advan1}
# TCW: TCA/TCT
tbl_dat$APOBEC_target <- tbl_dat$Var1 %in% c("TCA","TCT")
apobec_dat <- aggregate(Freq ~ APOBEC_target,tbl_dat,FUN=sum,na.rm=TRUE)
apobec_dat
```

---
## APOBEC activation in this sample?
```{r motif_ApoTar_advan2,eval=FALSE,echo=TRUE,tidy=FALSE,fig.align="center"}
ggplot(apobec_dat,aes(x=APOBEC_target,y=Freq,fill=APOBEC_target))+
  geom_bar(stat='identity')+
  labs(x="",y="Variants",fill="")+
  theme(legend.position = "none")
```

---
## APOBEC activation in this sample?
```{r motif_ApoTar_advan3,eval=TRUE,echo=FALSE,tidy=FALSE,fig.align="center"}
ggplot(apobec_dat,aes(x=APOBEC_target,y=Freq,fill=APOBEC_target))+
  geom_bar(stat='identity')+
  labs(x="",y="Variants",fill="")+
  theme(legend.position = "none")
```
***Probably not APOBEC-enriched***

---
## Exercise Time
- [Exercise](../../exercises/exercises/GenomicVariant_session1_exercise.html)
- [Answer](../../exercises/answers/GenomicVariant_session1_answers.html)